lib/ts/BufferWriterFormat.cc

```diff
@@ -60,7 +60,26 @@ tv_to_positive_decimal(ts::TextView src, ts::TextView *out)
 
 namespace ts
 {
-const ts::BWFSpec ts::BWFSpec::DEFAULT;
+const BWFSpec BWFSpec::DEFAULT;
+const uint8_t BWFSpec::TYPE_FLAG[0x100] = {
+  /* 0     1     2     3     4     5     6     7     8     9     A     B     C     D     E     F*/
+  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // 00
+  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // 10
+  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // 20
+  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // 30
+  0x00, 0x00, 0x07, 0x00, 0x07, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // 40
+  0x05, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x07, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // 50
+  0x00, 0x00, 0x03, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x03, // 60
+  0x01, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x03, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // 70
+  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // 80
+  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // 90
+  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // A0
+  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // B0
+  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // C0
+  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // D0
+  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // E0
+  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // F0
+};
```

Comment:
Correct me if I'm wrong, but it seems:
* `s` (`0x73`) doesn't appear to be valid, but it is set to `0x01`.
* `d` (`0x64`) is missing, and should be `0x03`.

Reply:
Is there any way we could make use of designated initializers from c99?
```C
const uint8_t BWFSpec::TYPE_FLAG[0x100] = {
  ['b'] = TYPE_FLAG_VALID | TYPE_FLAG_NUMERIC,
  ['B'] = TYPE_FLAG_VALID | TYPE_FLAG_NUMERIC | TYPE_FLAG_UPPER,
  ['o'] = TYPE_FLAG_VALID | TYPE_FLAG_NUMERIC,
  ['d'] = TYPE_FLAG_VALID | TYPE_FLAG_NUMERIC,
  ['D'] = TYPE_FLAG_VALID | TYPE_FLAG_NUMERIC | TYPE_FLAG_UPPER,
  ['x'] = TYPE_FLAG_VALID | TYPE_FLAG_NUMERIC,
  ['X'] = TYPE_FLAG_VALID | TYPE_FLAG_NUMERIC | TYPE_FLAG_UPPER,
  ['p'] = TYPE_FLAG_VALID,
  ['P'] = TYPE_FLAG_VALID                     | TYPE_FLAG_UPPER,
  ['g'] = TYPE_FLAG_VALID,
  ['s'] = TYPE_FLAG_VALID,
};
```

Or could we make a class with which we could use aggregate initialization?

The reason I am asking is because this block mostly looks like noise rather than telling me what our code is doing. It's already arranged in a way that helps readability, but I think so long as we don't make things too slow even more readability would help.

Reply:
Looks like the C99 style initialization doesn't work.
```
BufferWriterFormat.cc:85:67: sorry, unimplemented: non-trivial designated initializers not supported
 const uint8_t BWFSpec::TYPE_FLAG[0x100] = {['b'] = TYPE_FLAG_VALID};
```

Reply:
Perhaps later. I think it is doable but it would be quite a bit of work. That would also require some rather complex C++ technology on a PR on which there are already complaints on that subject.

Reply:
Here's a way of doing a table like this that will maybe cause @zwoop to loose the will to live:

http://coliru.stacked-crooked.com/a/1f6b320b35348dca

Reply:
@SolidWallOfCode How about something like this?

```Diff
diff --git a/lib/ts/BufferWriterFormat.cc b/lib/ts/BufferWriterFormat.cc
index 3d8709555..13fc08644 100644
--- a/lib/ts/BufferWriterFormat.cc
+++ b/lib/ts/BufferWriterFormat.cc
@@ -62,39 +62,6 @@ namespace ts
 {
 const BWFSpec BWFSpec::DEFAULT;
 
-/*
-  'B' : 0x42
-  'b' : 0x62
-  'd' : 0x74
-  'g' : 0x67
-  'o' : 0x6F
-  'p' : 0x50
-  'P' : 0x70
-  'S' : 0x53
-  's' : 0x73
-  'X' : 0x58
-  'x' : 0x78
- */
-const uint8_t BWFSpec::TYPE_FLAG[0x100] = {
-  /* 0     1     2     3     4     5     6     7     8     9     A     B     C     D     E     F*/
-  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // 00
-  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // 10
-  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // 20
-  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // 30
-  0x00, 0x00, 0x07, 0x00, 0x07, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // 40
-  0x05, 0x00, 0x00, 0x05, 0x00, 0x00, 0x00, 0x00, 0x07, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // 50
-  0x00, 0x00, 0x03, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x03, // 60
-  0x01, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x03, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // 70
-  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // 80
-  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // 90
-  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // A0
-  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // B0
-  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // C0
-  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // D0
-  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // E0
-  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // F0
-};
-
 /// Parse a format specification.
 BWFSpec::BWFSpec(TextView fmt)
 {
diff --git a/lib/ts/BufferWriterForward.h b/lib/ts/BufferWriterForward.h
index ab69f8d33..5d071ca89 100644
--- a/lib/ts/BufferWriterForward.h
+++ b/lib/ts/BufferWriterForward.h
@@ -79,6 +79,8 @@ struct BWFSpec {
   /// Check if the type in @a this is an upper case variant.
   bool has_upper_case_type() const;
 
+  class TypeFlag;
+
 protected:
   /// Validate character is alignment character and return the appropriate enum value.
   Align align_of(char c);
@@ -89,8 +91,6 @@ protected:
   static constexpr uint8_t TYPE_FLAG_VALID   = 0x01; ///< A valid type character.
   static constexpr uint8_t TYPE_FLAG_NUMERIC = 0x02; ///< Numeric output.
   static constexpr uint8_t TYPE_FLAG_UPPER   = 0x04; ///< Upper case flag.
-  /// Type character properties.
-  static const uint8_t TYPE_FLAG[0x100];
 };
 
 inline BWFSpec::Align
@@ -105,6 +105,35 @@ BWFSpec::is_sign(char c)
   return '+' == c || '-' == c || ' ' == c;
 }
 
+/// Type character properties.
+class BWFSpec::TypeFlag {
+  public:
+  TypeFlag() {
+    _type_flag['b'] = TYPE_FLAG_VALID | TYPE_FLAG_NUMERIC;
+    _type_flag['B'] = TYPE_FLAG_VALID | TYPE_FLAG_NUMERIC | TYPE_FLAG_UPPER;
+    _type_flag['o'] = TYPE_FLAG_VALID | TYPE_FLAG_NUMERIC;
+    _type_flag['d'] = TYPE_FLAG_VALID | TYPE_FLAG_NUMERIC;
+    _type_flag['D'] = TYPE_FLAG_VALID | TYPE_FLAG_NUMERIC | TYPE_FLAG_UPPER;
+    _type_flag['x'] = TYPE_FLAG_VALID | TYPE_FLAG_NUMERIC;
+    _type_flag['X'] = TYPE_FLAG_VALID | TYPE_FLAG_NUMERIC | TYPE_FLAG_UPPER;
+    _type_flag['p'] = TYPE_FLAG_VALID;
+    _type_flag['P'] = TYPE_FLAG_VALID                     | TYPE_FLAG_UPPER;
+    _type_flag['g'] = TYPE_FLAG_VALID;
+    _type_flag['s'] = TYPE_FLAG_VALID;
+  }
+  const uint8_t operator[](size_t index) const
+  {
+    ink_assert(index < _TYPE_FLAGS_SIZE);
+    return _type_flag[index];
+  }
+
+  private:
+  constexpr static size_t _TYPE_FLAGS_SIZE = 0x100;
+  uint8_t _type_flag[_TYPE_FLAGS_SIZE];
+};
+
+static const BWFSpec::TypeFlag TYPE_FLAG;
+
 inline bool
 BWFSpec::is_type(char c)
 {

```